### PR TITLE
[Test]Refactor test/test_reductions.py to be device generic

### PR DESCRIPTION
**Pull request by kumartanmay-28**
Replace @onlyOn(["cuda", "xpu"]) with @onlyAccelerator for 12 tests to enable execution on all accelerators including out of tree devices via PrivateUse1.

  Changes in test/test_reductions.py:

  Add import: onlyAccelerator from common_device_type
  Replace @onlyOn(["cuda", "xpu"]) with @onlyAccelerator for 12 tests:
  - test_mode_large, test_mode_wrong_device, test_prod_gpu
  - test_var, test_var_large_input, test_sum_noncontig
  - test_min_max_nan, test_sum_cpu_device_mismatch
  - test_reduction_split, test_reduction_vectorize_along_input_corner
  - test_argminmax_large_axis, test_histc_min_max_corner_cases_cuda
  Keep @onlyOn(["cuda", "xpu"]) for test_reductions_large_half_tensors (tests CUDA-specific mean_cuda limitation for ComplexHalf)
  Keep @onlyCUDA for test_reduction_vectorize_along_output (XPU driver issue noted in comments)

  Test Plan:
  Verified @onlyAccelerator test on CUDA: test_mode_wrong_device_cuda - OK
  Verified @onlyAccelerator test skipped on CPU: test_mode_wrong_device_cpu - OK (skipped)
  Verified @onlyAccelerator test on CUDA: test_prod_gpu_cuda_float16 - OK
  Verified @onlyAccelerator test skipped on CPU: test_prod_gpu_cpu_float16 - OK (skipped)
  Verified @onlyCUDA test on CUDA: test_reduction_vectorize_along_output_cuda_float32 - OK
  Verified @onlyCUDA test skipped on CPU: test_reduction_vectorize_along_output_cpu_float32 - OK (skipped)
  
  CC: @albanD @fffrog @jbschlosser

### CODE REVIEW

File: test/test_reductions.py
Comment (fffrog) [inline test/test_reductions.py]: Hi @kumartanmay-28, as indicated by the TODO comment, we need to pass a device parameter to _test_dim_ops so that it doesn't rely on the global variable defined on line 34.

Comment (fffrog) [inline test/test_reductions.py]: We need to eliminate this one and switch to using the device parameter rather than the global variable device_type. For example:

```
    @onlyCPU
    def test_max_mixed_devices(self, device):
        a = torch.randn(10, device=device)
        if torch.accelerator.is_available():
            values = torch.randn(10).to(device_type)
            indices = torch.LongTensor().to(device_type)
            self.assertRaises(RuntimeError,
                              lambda: torch.max(a, 0, out=(values, indices)))
            self.assertRaises(RuntimeError,
                              lambda: torch.amax(a, 0, out=values))
```

File: test/test_reductions.py [hunk 1]
```diff
@@ -1543,7 +1543,7 @@ def test_sum_out(self, device, dtype: torch.dtype) -> None:
         torch.sum(x, (2, 1), out=res2)
         self.assertEqual(res1, res2)
 
-    @onlyOn(["cuda", "xpu"])
+    @onlyAccelerator
     @dtypes(torch.float16, torch.float32)
     def test_prod_gpu(self, device, dtype):
```
Comment (fffrog) [inline test/test_reductions.py]: Make this name to be device-agnostic, please

File: test/test_reductions.py [hunk 2]
```diff
@@ -3297,7 +3297,7 @@ def test_histc_value_corner_cases(self, device, dtype):
             bins=4)
         self.assertEqual(3.0, actual.sum())
 
-    @onlyOn(["cuda", "xpu"])
+    @onlyAccelerator
     @dtypes(torch.uint8, torch.int8, torch.int, torch.long)
     def test_histc_min_max_corner_cases_cuda(self, device, dtype):
```
Comment (fffrog) [inline test/test_reductions.py]: ditto, making this name to be device-agnostic

File: test/test_reductions.py
Comment (fffrog) [inline test/test_reductions.py:2563]: Using `@onlyAccelerator` please, and if some in-tree accelerators fail, we can add `@skipIf(xxx)` for them.

Comment (fffrog) [inline test/test_reductions.py]: All the `onlyNativeDeviceTypes` can be replaced with `onlyAccelerator`

Comment (fffrog) [inline test/test_reductions.py]: This is very familar with `_test_dim_ops`, we need to use the parameter `device` rather than relying on the global one.

**Comment by fffrog** [comment]
BTW, most of `onlyCPU` can be removed and make those testcases to be runned on other accelerators.

**Comment by fffrog** [comment]
The changes are not that massive, so I think we can refactor this file thoroughly in a single PR. What do you think?

**Review by kumartanmay-28 — COMMENTED** [review]
Thanks for the detailed feedback @fffrog !

I have made the following changes as per your suggestions:

- Removed the global device_type variable  and updated all references accordingly.
- Updated the _test_dim_ops helper to accept a device parameter and use torch.accelerator.current_accelerator().type instead of the global device_type.
- Updated the _test_reduce_integer_upcast helper to accept a device parameter and pass it to torch.ones() and get_all_math_dtypes().
- Updated all test calls to helper functions so that test_sum_dim, test_prod_dim, test_mean_dim, test_std_dim, and test_var_dim now pass the device parameter.
- Renamed test_prod_gpu to test_prod_accelerator.
- Updated test_max_mixed_devices to replace the global device_type usage with torch.accelerator.current_accelerator().type.
- Renamed test_histc_min_max_corner_cases_cuda to test_histc_min_max_corner_cases_accelerator.
- Replaced all remaining decorators as noted above.


Verified the tests too:
CPU: 9,444 passed (112 expected failures from pre-existing OpInfo tests)
CUDA: 9,444 passed (112 expected failures from pre-existing OpInfo tests)

Thank you!

**Comment by kumartanmay-28** [comment]
> The changes are not that massive, so I think we can refactor this file thoroughly in a single PR. What do you think?

Agreed @fffrog ! I have made the changes accordingly in this PR itself! Thank you for the detailed review!

**Comment by kumartanmay-28** [comment]
> BTW, most of `onlyCPU` can be removed and make those testcases to be runned on other accelerators.

I’ll raise a follow up PR to replace the remaining ~17 @onlyCPU decorators that do not appear to have CPU specific logic. Please let me know if that approach works for you, @fffrog.

File: test/test_reductions.py [hunk 3]
```diff
@@ -844,8 +838,9 @@ def do_one(tensors_dict, dim):
                         expected = numpy_op(tensor.cpu().numpy(), dim)
                     actual = pytorch_op(tensor, dim)
                     self._assert_matches_numpy(actual, expected)
-                    if device_type in ["cuda", "xpu"]:
-                        self._assert_matches_numpy(pytorch_op(tensor.to(device_type), dim).cpu(), expected)
+                    if torch.accelerator.is_available():
+                        accelerator_type = torch.accelerator.current_accelerator().type
+                        self._assert_matches_numpy(pytorch_op(tensor.to(accelerator_type), dim).cpu(), expected)
```
Comment (fffrog) [inline test/test_reductions.py:843]: Hi, I think we can just retain this line: `self._assert_matches_numpy(pytorch_op(tensor.to(device_type), dim).cpu(), expected)` and remove the other device checks. Any device-specific restrictions can be handled within their respective test cases.

File: test/test_reductions.py
Comment (fffrog) [inline test/test_reductions.py:1593]: `@onlyAccelerator`

File: test/test_reductions.py [hunk 4]
```diff
@@ -1595,8 +1594,9 @@ def test_prod_bool(self, device):
     def test_max_mixed_devices(self, device):
         a = torch.randn(10, device=device)
```
Comment (fffrog) [inline test/test_reductions.py]:
```suggestion
        a = torch.randn(10, device="cpu")
```

File: test/test_reductions.py [hunk 5]
```diff
@@ -1595,8 +1594,9 @@ def test_prod_bool(self, device):
...
         if torch.accelerator.is_available():
-            values = torch.randn(10).to(device_type)
-            indices = torch.LongTensor().to(device_type)
+            accelerator_type = torch.accelerator.current_accelerator().type
+            values = torch.randn(10).to(accelerator_type)
+            indices = torch.LongTensor().to(accelerator_type)
```
Comment (fffrog) [inline test/test_reductions.py:1599]:
```
values = torch.randn(10).to(device)
indices = torch.LongTensor().to(device)
```

File: test/test_reductions.py
Comment (fffrog) [inline test/test_reductions.py]:
```suggestion
    @onlyAccelerator
```

File: test/test_reductions.py [hunk 6]
```diff
@@ -1606,8 +1606,9 @@ def test_max_mixed_devices(self, device):
     def test_min_mixed_devices(self, device):
         a = torch.randn(10, device=device)
```
Comment (fffrog) [inline test/test_reductions.py]:
```suggestion
        a = torch.randn(10, device="cpu")
```

File: test/test_reductions.py [hunk 7]
```diff
@@ -1606,8 +1606,9 @@ def test_max_mixed_devices(self, device):
...
         if torch.accelerator.is_available():
-            values = torch.randn(10).to(device_type)
-            indices = torch.LongTensor().to(device_type)
+            accelerator_type = torch.accelerator.current_accelerator().type
+            values = torch.randn(10).to(accelerator_type)
+            indices = torch.LongTensor().to(accelerator_type)
```
Comment (fffrog) [inline test/test_reductions.py:1611]:
```
values = torch.randn(10).to(device)
indices = torch.LongTensor().to(device)
```

File: test/test_reductions.py [hunk 8]
```diff
@@ -1096,7 +1095,7 @@ def test_mode_boolean(self, device):
 
 
     @expectedFailureMeta  # mode only supports CPU and CUDA device type
-    @onlyNativeDeviceTypes
```
Comment (fffrog) [inline test/test_reductions.py:1099]: Delete `onlyNativeDeviceTypes` directly and do not use extra `onlyAccelerator` to keep the consistent with the original code.

**Comment by can-gaa-hou** [comment]
closed with a similar merged pr https://github.com/pytorch/pytorch/pull/185881
